platform/MCU/nRF52840/src/flash_mal.c

```diff
@@ -368,23 +368,31 @@ int FLASH_CopyMemory(flash_device_t sourceDeviceID, uint32_t sourceAddress,
         return FLASH_ACCESS_RESULT_BADARG;
     }
 #if SOFTDEVICE_MBR_UPDATES
-    if (module_function == MODULE_FUNCTION_BOOTLOADER && destinationAddress == USER_FIRMWARE_IMAGE_LOCATION) {
+    if (module_function == MODULE_FUNCTION_BOOTLOADER && destinationAddress == USER_FIRMWARE_IMAGE_LOCATION_COMPAT) {
         // Backup user firmware
         dest_size = CEIL_DIV(length, INTERNAL_FLASH_PAGE_SIZE) * INTERNAL_FLASH_PAGE_SIZE;
         if (!FLASH_EraseMemory(FLASH_SERIAL, EXTERNAL_FLASH_RESERVED_ADDRESS, dest_size)) {
             return FLASH_ACCESS_RESULT_ERROR;
         }
-        if (!flash_copy(FLASH_INTERNAL, USER_FIRMWARE_IMAGE_LOCATION, FLASH_SERIAL, EXTERNAL_FLASH_RESERVED_ADDRESS, dest_size)) {
+        if (!flash_copy(FLASH_INTERNAL, USER_FIRMWARE_IMAGE_LOCATION_COMPAT, FLASH_SERIAL, EXTERNAL_FLASH_RESERVED_ADDRESS, dest_size)) {
```

Comment:
If we are running the application@0xb4000 and its size is larger than 128KB, we need to backup it instead?

Reply:
We don't need to do that. This change simply keeps the existing behavior where the part of the 0xd40000 application (and now potentially 0xb40000 application) is temporarily used for storing the bootloader binary for MBR update. For all intents and purposes it does not matter where we take the chunk out of temporarily (it's anyway going to be restored after the bootloader update is applied). This is just specifically kept at the same locations as < 3.1.0 Device OS "just in case".